core/src/main/scala/org/genivi/sota/core/db/Campaigns.scala

```diff
@@ -0,0 +1,199 @@
+/**
+ * Copyright: Copyright (C) 2016, ATS Advanced Telematic Systems GmbH
+ * License: MPL-2.0
+ */
+package org.genivi.sota.core.db
+
+import eu.timepit.refined.refineV
+import java.time.Instant
+import java.util.UUID
+import org.genivi.sota.core.data.Campaign
+import org.genivi.sota.core.SotaCoreErrors
+import org.genivi.sota.data.{GroupInfo, Namespace, PackageId}
+import org.genivi.sota.db.SlickExtensions._
+import scala.concurrent.ExecutionContext
+import scala.util.{Failure, Success}
+import slick.driver.MySQLDriver.api._
+
+object Campaigns {
+  import org.genivi.sota.refined.SlickRefined._
+  import Campaign._
+
+  // scalastyle:off
+  class CampaignTable(tag: Tag) extends Table[CampaignMeta](tag, "Campaign") {
+    def id = column[Id]("uuid")
+    def namespace = column[Namespace]("namespace")
+    def name = column[String]("name")
+    def launched = column[Boolean]("launched")
+    def packageName = column[PackageId.Name]("package_name")
+    def packageVersion = column[PackageId.Version]("package_version")
+
+    def pk = primaryKey("pk_campaign", id)
+
+    private def mPackageId(name0: Option[PackageId.Name],
+                           version0: Option[PackageId.Version]): Option[PackageId] = for
+      { name <- name0
+        version <- version0
+      } yield PackageId(name, version)
+
+    def * = (id, namespace, name, launched, packageName.?, packageVersion.?).shaped <>
+      ({case (id,ns, name, launched, pkgName, pkgVer) =>
+        CampaignMeta(id,ns, name, launched, mPackageId(pkgName, pkgVer))},
+       (x: CampaignMeta) => Some ((x.id, x.namespace, x.name, x.launched,
+                                   x.packageId.map(_.name), x.packageId.map(_.version))))
+  }
+
+  class CampaignGroups(tag: Tag) extends Table[(Campaign.Id, GroupInfo.Name)](tag, "CampaignGroups") {
+    def campaignId = column[Campaign.Id]("campaign_id")
+    def group = column[GroupInfo.Name]("group_name")
+
+    def pk = primaryKey("pk_campaign_group", (campaignId, group))
+
+    def * = (campaignId, group)
+
+  }
+  // scalastyle:on
+
+  val campaignsMeta = TableQuery[CampaignTable]
+  val campaignsGroups = TableQuery[CampaignGroups]
+
+  def canEdit(ns: Namespace, id: Id)
+             (implicit ec: ExecutionContext): DBIO[CampaignMeta] =
+    exists(ns, id).flatMap { camp =>
+      if (camp.launched){ DBIO.failed(SotaCoreErrors.CampaignLaunched) }
+       else {DBIO.successful(camp) }
+    }
+
+  def create(ns: Namespace, name: String)
+            (implicit ec: ExecutionContext): DBIO[Campaign.Id] = {
+    val id: Id = Id(refineV[ValidId](UUID.randomUUID.toString).right.get)
+
+    val dbIO = for {
+      _ <- exists(id).asTry.flatMap {
```

Comment:
we should not have this explicitly but try to insert into the db and handle the integrity error (see `handleIntegrityErrors`)